### PR TITLE
RFC: Style and shellchecks

### PR DESCRIPTION
**Pull request by zimbatm**
###### Things done:
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [ ] Built on platform(s): NixOS / OSX / Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Using ShellCheck to make sure all of our bash is escaped properly. Is this something desirable ?

There are a couple of things I see as controversial in these changes:

Use `#!bash` is a hack that lets vim and shellcheck know that we are dealing with a bash dialect. The file itself is not executable because a shebang needs an absolute path (thus /usr/bin/env's existance).

Use `[[ ]]` over the POSIX `[ ]` or `test` syntax. Mainly because it's smarter in regards to escaping and less likely to mis-use.

A new `--unset` option that I bundle with these changes because of the mass-rebuild. 

---

This is not done yet, just seeking for confirmation if this is a good idea or not. Also trying to see if we can agree on a bash style.

/cc @edolstra @domenkozar

### CODE REVIEW

File: pkgs/build-support/setup-hooks/make-wrapper.sh [hunk 1]
```diff
@@ -1,86 +1,99 @@
...
+            exit 1
+        esac
     done
 
     # Note: extraFlagsArray is an array containing additional flags
     # that may be set by --run actions.
-    echo exec ${argv0:+-a $argv0} "$original" \
-         $flagsBefore '"${extraFlagsArray[@]}"' '"$@"' >> $wrapper
+    echo "exec ${argv0:+-a $argv0 }$original" \
+         "$flagsBefore" '"${extraFlagsArray[@]}"' '"$@"' >> "$wrapper"
```
Comment (zimbatm) [inline pkgs/build-support/setup-hooks/make-wrapper.sh]: I need to fix the escaping of $flagsBefore here

File: pkgs/stdenv/generic/setup.sh [hunk 2]
```diff
@@ -16,8 +17,8 @@ runHook() {
     local hookName="$1"
     shift
     local var="$hookName"
-    if [[ "$hookName" =~ Hook$ ]]; then var+=s; else var+=Hooks; fi
-    eval "local -a dummy=(\"\${$var[@]}\")"
+    if [[ $hookName =~ Hook$ ]]; then var+=s; else var+=Hooks; fi
+    eval "local -a dummy=(\"\${${var[*]}}\")"
```
Comment (zimbatm) [inline pkgs/stdenv/generic/setup.sh]: This-one is tricky and needs to be tested properly.

**Comment by ttuegel** [comment]
These are mass-rebuilding changes; please make this PR against the staging branch.

**Comment by zimbatm** [comment]
Okay. Should I re-open another PR ? I don't think github allows me to switch the target.

**Comment by edolstra** [comment]
Looks good. Didn't know that quotes are unnecessary in `[[ ... ]]` tests.

**Comment by dezgeg** [comment]
> Use `[[ ]]` over the POSIX `[ ]` or test syntax. Mainly because it's smarter in regards to escaping and less likely to mis-use.

IMHO `[[ ]]` is not better - you need to learn *totally* new quoting rules for that, e.g.
d844a1002f083d81084203f8f3431421a0fc1698

**Comment by ttuegel** [comment]
> Should I re-open another PR ? I don't think github allows me to switch the target.

Yes. Sorry!

**Comment by zimbatm** [comment]
Alright, everyone to the batmobile ! #13543

**Pull request closed without merging**